### PR TITLE
[Dataset] Basic template for DGL Dataset

### PR DESCRIPTION
**Pull request by classicsong**
## Description
Basic template for DGL Dataset

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/data/dgl_dataset.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
+"""Basic DGL Dataset
+"""
+
+from __future__ import absolute_import
+
+import os, sys
+from .utils import download, extract_archive, get_download_dir, makedirs
+from ..utils import retry_method_with_fix
+
+class DGLDataset(object):
```
Comment (hetong007) [inline python/dgl/data/dgl_dataset.py:10]: Will we introduce the `F.dataset` base class now or in the later version?

File: python/dgl/data/dgl_dataset.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+        self._load()
+
+    def download(self):
+        r"""Downloads the dataset to the :obj:`self.raw_dir` folder.
+            Can be ignored if the dataset is already in self.raw_dir
+        """
+        pass
+
+    def save(selfs):
```
Comment (VoVAllen) [inline python/dgl/data/dgl_dataset.py:44]: self?

File: python/dgl/data/dgl_dataset.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        Builtin dataset will be automatically downloaded into ~/.dgl/
+    Parameters
+    name : str
+        Name of the dataset
+    url : str
+        Url to download the raw dataset
+    force_reload : bool
+        Whether to reload the dataset. Default: False
+    """
+    def __init__(self, name, url, force_reload=False):
```
Comment (hetong007) [inline python/dgl/data/dgl_dataset.py:145]: Since `force_reload=False` here, I suppose all sub-classes in the API doc can get rid of it?

File: python/dgl/data/dgl_dataset.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
+"""Basic DGL Dataset
+"""
+
+from __future__ import absolute_import
+
+import os, sys
+from .utils import download, extract_archive, get_download_dir, makedirs
+from ..utils import retry_method_with_fix
```
Comment (hetong007) [inline python/dgl/data/dgl_dataset.py:8]: I don't see anywhere other than `dgl.data` using `retry_method_with_fix`. Do we consider move it into `dgl.data`?

File: python/dgl/data/dgl_dataset.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    def download(self):
+        r"""Downloads the dataset to the :obj:`self.raw_dir` folder.
+            Can be ignored if the dataset is already in self.raw_dir
+        """
+        pass
+
+    def save(selfs):
+        r"""Save the processed dataset into files.
+            Use dgl.utils.data.save_graphs to save dgl graph into files.
+            Use dgl.utils.data.save_info to save extra dict information into files.
```
Comment (hetong007) [inline python/dgl/data/dgl_dataset.py:47]: What are the typical `extra dict info` here? Graph classification may have graph-level labels that can't be included in `ndata`/`edata`.

File: python/dgl/data/dgl_dataset.py (hunk 1 above)
Comment (classicsong) [reply python/dgl/data/dgl_dataset.py:10]: Maybe later

File: python/dgl/data/dgl_dataset.py (hunk 3 above)
Comment (classicsong) [reply python/dgl/data/dgl_dataset.py:145]: Yes

File: python/dgl/data/dgl_dataset.py (hunk 5 above)
Comment (classicsong) [reply python/dgl/data/dgl_dataset.py:47]: labels, for heterograph, there maybe target type, etc.

Comment (hetong007) [reply python/dgl/data/dgl_dataset.py:47]: Then are we gonna save/load the graph labels with another function?

File: python/dgl/data/dgl_dataset.py (hunk 2 above)
Comment (classicsong) [reply python/dgl/data/dgl_dataset.py]: yes

File: python/dgl/data/dgl_dataset.py (hunk 4 above)
Comment (classicsong) [reply python/dgl/data/dgl_dataset.py:8]: You can do it in the next PR.

**Comment by VoVAllen** [comment]
What are the minimal methods needed to implement for a new dataset?

**Comment by VoVAllen** [comment]
Can we make it as declarative as possible as this example https://github.com/pytorch/vision/blob/master/torchvision/datasets/cifar.py#L142-L162?

**Comment by classicsong** [comment]
> What are the minimal methods needed to implement for a new dataset?
The minimal is process().
The suggested way is to implement save(), load() and has_cache() as well.

**Comment by classicsong** [comment]
> Can we make it as declarative as possible as this example https://github.com/pytorch/vision/blob/master/torchvision/datasets/cifar.py#L142-L162?
declarative loader will be another task: dgl.dataloader. We need to postpone that after 0.5 release.
